### PR TITLE
Add API to get commit diff/patch

### LINKED ISSUE #10921: [Feature] [API] GET commit content
this api does not exist on Github and its needed for web clients to view commit content (they usuale dont have git installed :wink: )

### PR DESCRIPTION
**Pull request by qwerty287**
This PR adds an API endpoint to get the diff or patch file of a commit. It uses the same mechanism as the "site" `<host>/<owner>/<repo>/commit/<sha>.<diff or patch>`.

Endpoints:
Get diff `/repos/{owner}/{repo}/git/commits/{sha}.diff`
Get patch `/repos/{owner}/{repo}/git/commits/{sha}.patch`

Could close #10921 #13012

### CODE REVIEW

File: routers/api/v1/api.go [hunk 1]
```diff
@@ -937,6 +937,10 @@ func Routes(sessioner func(http.Handler) http.Handler) *web.Route {
 				m.Group("/git", func() {
 					m.Group("/commits", func() {
 						m.Get("/{sha}", repo.GetSingleCommit)
+						m.Get("/{sha}.diff",
```
Comment (6543) [inline routers/api/v1/api.go]:
```suggestion
						m.Get("/{sha}.{diffType:diff|patch}",
```
can we unify that function directly within the router - so we only have one new endpoint

File: routers/api/v1/repo/commits.go [hunk 2]
```diff
@@ -213,3 +213,83 @@ func GetAllCommits(ctx *context.APIContext) {
...
+	//   in: path
+	//   description: name of the repo
+	//   type: string
+	//   required: true
+	// - name: sha
+	//   in: path
+	//   description: SHA of the commit to get
+	//   type: string
+	//   required: true
+	// responses:
```
Comment (6543) [inline routers/api/v1/repo/commits.go]:
```suggestion
	// - name: diffType
	//   in: path
	//   type: string
	...
	// responses:
```

Comment (6543) [reply routers/api/v1/repo/commits.go]: (needed for unify)

File: routers/api/v1/api.go (hunk 1 above)
Comment (qwerty287) [reply routers/api/v1/api.go]: I'll look into this, just wait a moment 🙃

Comment (qwerty287) [reply routers/api/v1/api.go]: I merged them and it works... But `make generate-swagger` doesn't work. It always returns (the swagger JSON in the commit is something I tried to resolve it, but it's not that what it should be):
```
operation (repoDownloadCommitDiffOrPatch): yaml: line 21: did not find expected key
exit status 1
```

Comment (qwerty287) [reply routers/api/v1/api.go]: That's also the reason why the CI fails

Comment (6543) [reply routers/api/v1/api.go]: use space not tabs

Comment (qwerty287) [reply routers/api/v1/api.go]: Thanks for solving this :)

**Comment by 6543** [comment]
could you write a very simple integration test (one download of a patch and one for diff) ...
If you need help just ask

**Comment by qwerty287** [comment]
Is https://github.com/go-gitea/gitea/pull/17095/commits/67ebe5070d63b41e825f609446598459f6b46b09 enough?

**Review by zeripath — APPROVED** [review]

**Review by axifive — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Comment by qwerty287** [comment]
Just a question, nothing that should be done in this PR: as @6543 suggested, I merged the two endpoints into one. Is there a certain reason why there are two endpoints for the PRs? One for diff and one for patch? To keep it consistent, it may be good to merge them into one.

**Comment by axifive** [comment]
@qwerty287 as I can see they can be merged too

**Comment by 6543** [comment]
> Just a question, nothing that should be done in this PR: as @6543 suggested, I merged the two endpoints into one. Is there a certain reason why there are two endpoints for the PRs? One for diff and one for patch? To keep it consistent, it may be good to merge them into one.

it would be breaking in terms of swagger doc but not for api - so I would say it's good to go

yes if you would open another pull I'm happy to look at it

**Comment by qwerty287** [comment]
Ok I think I'd like to do this

**Pull request merged**